src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,399 @@
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/config.hpp" // for TORRENT_USE_RTC
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/websocket_tracker_connection.hpp"
+#include "libtorrent/aux_/escape_string.hpp"
+#include "libtorrent/aux_/session_settings.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/ip_filter.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/socket_io.hpp"
+#include "libtorrent/tracker_manager.hpp"
+
+#include "nlohmann/json.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio> // for snprintf
+#include <functional>
+#include <list>
+#include <locale>
+#include <string>
+#include <vector>
+
+namespace libtorrent {
+
+using namespace std::placeholders;
+
+using json = nlohmann::json;
+
+std::string from_latin1(std::string const& s) {
+	// Convert ISO-8859-1 (aka latin1) input to UTF-8
+	std::string r;
+	r.reserve(s.size()*2);
+	for(unsigned char c : s)
+	{
+		if (!(c & 0x80))
+		{
+			r.push_back(c);
+		}
+		else {
+			// 2 bytes needed for code point
+			r.push_back(0xC0 | (c >> 6));
+			r.push_back(0x80 | (c & 0x3F));
+		}
+	}
+	return r;
+}
+
+std::string to_latin1(std::string const& s) {
+	std::string r;
+	r.reserve(s.size());
+	auto it = s.begin();
+	while(it != s.end())
+	{
+		unsigned char c = *it;
+
+		// Find the number of bytes
+		size_t len;
+		if(!(c & 0x80)) len = 1;
+		else if (!(c & 0x20)) len = 2;
+		else if (!(c & 0x10)) len = 3;
+		else if (!(c & 0x08)) len = 4;
+		else if (!(c & 0x04)) len = 5;
+		else len = 6;
```

Comment:
any length > 2 should be an error here, right?

Reply:
Indeed it shouldn't happen here, I handled longer lengths only to have a consistent UTF-8 to latin1 conversion.